sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -149,14 +149,144 @@ trait CodegenSupport extends SparkPlan {
+   * we may put the consume logic of parent operator into a function and set this flag to `true`.
+   * The parent operator can know if its consume logic is inlined or in separated function.
+   */
+  private var doConsumeInFunc: Boolean = false
+
+  /**
+   * Returning true means we have at least one consume logic from child operator or this operator is
+   * separated in a function. If this is `true`, this operator shouldn't use `continue` statement to
+   * continue on next row, because its generated codes aren't enclosed in main while-loop.
+   *
+   * For example, we have generated codes for a query plan like:
+   *   Op1Exec
+   *     Op2Exec
+   *       Op3Exec
+   *
+   * If we put the consume code of Op2Exec into a separated function, the generated codes are like:
+   *   while (...) {
+   *     ... // logic of Op3Exec.
+   *     Op2Exec_doConsume(...);
+   *   }
+   *   private boolean Op2Exec_doConsume(...) {
+   *     ... // logic of Op2Exec to consume rows.
+   *   }
+   * For now, `doConsumeInChainOfFunc` of Op2Exec will be `true`.
+   *
+   * Notice for some operators like `HashAggregateExec`, it doesn't chain previous consume functions
+   * but begins with its produce framework. We should override `doConsumeInChainOfFunc` to return
+   * `false`.
+   */
+  protected def doConsumeInChainOfFunc: Boolean = {
+    val codegenChildren = children.map(_.asInstanceOf[CodegenSupport])
+    doConsumeInFunc || codegenChildren.exists(_.doConsumeInChainOfFunc)
+  }
+
+  /**
+   * The actual java statement this operator should use if there is a need to continue on next row
+   * in its `doConsume` codes.
+   *
+   *   while (...) {
+   *     ...       // logic of Op3Exec.
+   *     Op2Exec_doConsume(...);
+   *   }
+   *   private boolean Op2Exec_doConsume(...) {
+   *     ...       // logic of Op2Exec to consume rows.
+   *     continue; // Wrong. We can't use continue with the while-loop.
+   *   }
+   * In above code, we can't use `continue` in `Op2Exec_doConsume`.
+   *
+   * Instead, we do something like:
+   *   while (...) {
+   *     ...          // logic of Op3Exec.
+   *     boolean continueForLoop = Op2Exec_doConsume(...);
+   *     if (continueForLoop) continue;
+   *   }
+   *   private boolean Op2Exec_doConsume(...) {
+   *     ...          // logic of Op2Exec to consume rows.
+   *     return true; // When we need to do continue, we return true.
+   *   }
+   */
+  protected def continueStatementInDoConsume: String = if (doConsumeInChainOfFunc) {
+    "return true;";
+  } else {
+    "continue;"
+  }
+
+  /**
+   * To prevent concatenated function growing too long to be optimized by JIT. We can separate the
+   * parent's `doConsume` codes of a `CodegenSupport` operator into a function to call.
+   */
+  protected def constructDoConsumeFunction(
+      ctx: CodegenContext,
+      inputVars: Seq[ExprCode]): String = {
+    val (callingParams, arguList, inputVarsInFunc) =
```

Comment:
It would be good to add fall back path to the original code gen (i.e. without creating consume function) if the number of arguments is more than 254 (255 - one for non-static method) (e.g. #19082).
If the total number of arguments is more than 255, janino will fail its compilation.

Reply:
Good suggestion. I'll follow it.